src/dmd/root/array.d

```diff
@@ -296,6 +301,38 @@ public:
         return data[a .. b];
     }
 
+    /**
+     * Sort the elements of an array
+     *
+     * This function relies on `qsort`.
+     *
+     * Params:
+     *   pred = Predicate to use. Should be a function of type
+     *          `int function(scope const T*  e1, scope const T* e2) nothrow`.
+     *          The return value of this function should follow the
+     *          usual C rule: `e1 >= e2 ? (e1 > e2) : -1`.
+     *          The function can have D linkage.
+     *
+     * Returns:
+     *   A reference to this, for easy chaining.
+     */
+    extern(D) ref typeof(this) sort (alias pred) () nothrow
+    {
+        if (this.length < 2)
+            return this;
+        qsort(this.data.ptr, this.length, T.sizeof, &arraySortWrapper!(T, pred));
+        return this;
+    }
+
+    static if (is(typeof(this.data[0].opCmp(this.data[1])) : int))
```

Comment:
It would appear that this line triggers a silent forward reference bug when compiling frontend modules separately.  The early forced semantic of `typeof(this.data[0])` (a ClassDeclaration when compiling mtype.d) messes up the vtable calculation, and the offset is wrong (when calling isBaseOf).

This happens in at least the dmd-cxx implementation, a suitable reduction and testing 2.066 will need to be done as well.

Reply:
Reduced, any compiler older than 2.085.1 runs the risk of producing a corrupted compiler if you are using incremental/one module at a time compilation strategy (2.085.0 is the last broken release).

FYI @kinke, or ping any other LDC devs so you don't repeat the time I spent looking at this issue.  I've already committed a fix to dmd-cxx branch, see the cross-referenced merged PR.

Reply:
@ibuclaw :
Could you check if using a template constraint would work around the problem ? And if not, we can move the `static assert` in the body.

Reply:
I think template constraints would work just fine.  The reduced test I added reports the correct offset at least...

Reply:
However, on further testing even if we fix the problematic static if, you cannot actually _use_ this function anyway, as it relies on dual-context pointers, which is neither implemented in gdc nor ldc.

Reply:
I don't see where it does ? Let's move the discussiong to #10930 though.